qutip/fileio.py

```diff
@@ -217,15 +216,18 @@ def qsave(data, name='qutip_data'):
     ----------
     data : instance/array_like
         Input Python object to be stored.
-    filename : str
+    filename : str or pathlib.Path
         Name of output data file.
 
     """
     # open the file for writing
-    fileObject = open(name + '.qu', 'wb')
-    # this writes the object a to the file named 'filename.qu'
-    pickle.dump(data, fileObject)
-    fileObject.close()
+    file = Path(name)
+    if file.suffix != '.qu':
+        file.with_suffix('.qu')
```

Comment:
Previously the code always added `.qu` regardless -- e.g. `foo.qu` would be saved as `foo.qu.qu` and `foo.txt` would be saved as `foo.txt.qu`. The new code saves `foo.qu` as `foo.qu` and `foo.txt` as `foo.qu`.

I'm not a fan of the old behaviour particularly, but the new behaviour of replacing the suffix isn't much better -- both are unnecessary complications that just make life harder for QuTiP users. Typing `qsave("foo.qu")` is more explicit and is less for the user to remember about the underlying function works.

I would suggest we keep the old behaviour for 4.7 (i.e. in this PR) -- e.g. `file = file.with_suffix(file.suffix + ".qu")`.

If you are interested, you could then open up a follow up PR for QuTiP 5 (the `dev.major` branch) and remove the suffix addition entirely there.

Reply:
I have brought back the functionality and will open a pull request for v5.